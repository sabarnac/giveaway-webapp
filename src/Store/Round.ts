import { observable, computed } from "mobx";
import Participant from "./Participant";
import Config from "./Config/Config";
import chunk from "lodash.chunk";
import Match from "./Match";

/**
 * Class containing the details of a round of the tournament.
 */
export default class Round {
  @observable public matches: Match[];

  private _config: Config;

  private _shuffleParticipants = (participants: Participant[]): Participant[] =>
    this._config.randomGenerator.shuffle(participants);

  private _createMatch = (participants: Participant[]): Match =>
    new Match(this._config, participants);

  private _getMatches = (participants: Participant[]): Match[] =>
    chunk(
      this._shuffleParticipants(participants),
      this._config.participantsPerMatch,
    ).map(this._createMatch);

  public constructor(config: Config, participants: Participant[]) {
    this._config = config;

    this.matches = this._getMatches(participants);
  }

  private _getMatchParticipants = (match: Match): Participant[] =>
    match.participants;

  private _getMatchWinner = (match: Match): Participant => match.winner;

  private _getMatchLosers = (match: Match): Participant[] => match.losers;

  /**
   * Get the list of participants in the round.
   * @returns {Participant[]} The list of participants.
   */
  @computed public get participants(): Participant[] {
    return this.matches.map(this._getMatchParticipants).flat();
  }

  /**
   * Get the list of winners in the round.
   * @returns {Participant[]} The list of winners.
   */
  @computed public get winners(): Participant[] {
    return this.matches.map(this._getMatchWinner);
  }

  /**
   * Get the list of losers in the round.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this.matches.map(this._getMatchLosers).flat();
  }
}
