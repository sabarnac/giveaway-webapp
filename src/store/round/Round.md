Usage:

```tsx
import Match from "./store/round/match/Match";
import Participant from "./store/round/match/participant/Participant";

const round: Round = new Round(Config.instance, [
  new Participant("FOO"),
  new Participant("BAR")
]);
const roundId: Round = round.id; // "round-1"
const matches: Match[] = round.matches; // [Match("match-1"), ..., Match("match-x")]
const firstMatch: Match = round.firstMatch; // Match("match-1")
const lastMatch: Match = round.lastMatch; // Match("match-x")
const participants: Participant[] = round.participants; // [Participant("FOO"), Participant("BAR")]
const winners: Participant[] = round.winners; // [Participant("FOO")]
const losers: Participant[] = round.losers; // [Participant("BAR")]
```
