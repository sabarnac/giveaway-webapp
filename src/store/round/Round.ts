import { observable, computed } from "mobx";
import Participant from "./match/participant/Participant";
import Config from "../config/Config";
import chunk from "lodash.chunk";
import Match from "./match/Match";
import last from "lodash.last";
import RandomGenerator from "../config/RandomGenerator";

/**
 * Class containing the details of a round of the tournament.
 */
export default class Round {
  @observable private _id: string;
  @observable private _matches: Match[];

  private static counter: number = 1;

  private _config: Config;

  private _shuffleParticipants = (participants: Participant[]): Participant[] =>
    RandomGenerator.shuffle(participants);

  private _createMatch = (participants: Participant[]): Match =>
    new Match(participants);

  private _getMatches = (participants: Participant[]): Match[] =>
    chunk(
      this._shuffleParticipants(participants),
      this._config.participantsPerMatch
    ).map(this._createMatch);

  public constructor(config: Config, participants: Participant[]) {
    this._config = config;

    this._id = `round-${Round.counter++}`;
    this._matches = this._getMatches(participants);
  }

  private _getMatchParticipants = (match: Match): Participant[] =>
    match.participants;

  private _getMatchWinner = (match: Match): Participant => match.winner;

  private _getMatchLosers = (match: Match): Participant[] => match.losers;

  /**
   * Get the match ID.
   * @returns {string} The unique ID of the match.
   */
  @computed public get id(): string {
    return this._id;
  }

  /**
   * Get the list of matches in the round.
   * @returns {Match[]} The list of matches.
   */
  @computed public get matches(): Match[] {
    return this._matches;
  }

  /**
   * Get the first match of the round.
   * @returns {Match} The first match.
   */
  @computed public get firstMatch(): Match {
    return this._matches[0];
  }

  /**
   * Get the last match of the round.
   * @returns {Match} The last match.
   */
  @computed public get lastMatch(): Match {
    return <Match>last(this._matches);
  }

  /**
   * Get the list of participants in the round.
   * @returns {Participant[]} The list of participants.
   */
  @computed public get participants(): Participant[] {
    return this._matches.map(this._getMatchParticipants).flat();
  }

  /**
   * Get the list of winners in the round.
   * @returns {Participant[]} The list of winners.
   */
  @computed public get winners(): Participant[] {
    return this._matches.map(this._getMatchWinner);
  }

  /**
   * Get the list of losers in the round.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this._matches.map(this._getMatchLosers).flat();
  }
}
