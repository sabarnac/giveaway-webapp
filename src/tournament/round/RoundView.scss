.round {
    width: 100%;
    height: 100%;
    opacity: 0.99999;
    transition: opacity 500ms ease-in-out;
    will-change: transition;
    &--exiting,
    &--exited {
        opacity: 0.000001;
    }
    &--entered,
    &--entering,
    &--completed {
        opacity: 0.99999;
    }
}