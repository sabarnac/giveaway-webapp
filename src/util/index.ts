export let isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

export let createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

export let isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;
