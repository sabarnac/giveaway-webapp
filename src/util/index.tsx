import React, { ComponentType } from "react";
import { Dispatch, useState } from "react";
import Config from "../store/config/Config";
import { Redirect } from "react-router";
import { Observer } from "mobx-react";

/**
 * Returns if it is currently a dev environment.
 * @return {boolean} Whether it is currently a dev environment.
 */
export const isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

/**
 * Returns a random avatar image URL.
 * @param {string} identifier The identifier to generate the avatar URL against.
 * @return {string} The avatar URL.
 */
export const createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

/**
 * Checks if the number is within the range.
 * @param {number} num The number to check.
 * @param {number} start The start of the range.
 * @param {number} end The end of the range.
 * @return {boolean} Whether the number is within the range.
 */
export const isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;

/**
 * Type for custom animation state hook.
 */
export type AnimationStateHook = [
  number,
  () => void,
  (delay: number) => void,
  (state: number) => void
];

/**
 * Type for show overlay state hook.
 */
export type ShowOverlayHook = [boolean, (newValue: boolean) => void];

/**
 * Type for void function.
 */
export type VoidFunction = () => void;

/**
 * Runs the given action at least after the given delay.
 * @param {Function} action The action to run.
 * @param {number} delay The minimum delay to wait for.
 */
export const runOnDelay = (action: Function, delay: number): number =>
  setTimeout(action, delay);

/**
 * Create an animation state hook.
 * @return {AnimationStateHook} The current animation state, a state update method, and a delayed state update method.
 */
export const useAnimationState = (start: number = 0): AnimationStateHook => {
  const [currentState, setState]: [number, Dispatch<number>] = useState(start);

  const updateState = (): void => setState(currentState + 1);
  const updateStateDelay = (delay: number): VoidFunction => {
    const delayId: number = runOnDelay(updateState, delay);
    return () => clearTimeout(delayId);
  };

  return [currentState, updateState, updateStateDelay, setState];
};

/**
 * Normalize the given animation time by the animation speed multiplier.
 * @param {number} time The animation time to normalize.
 * @return {number} The normalized animation speed
 */
export const getNormalizedSpeed = (time: number): number =>
  time / Config.getInstance().speed;

/**
 * Runs the given action if the predicate is true.
 * @param {boolean} predicate The predicate to check.
 * @param {Function} action The action to run.
 */
export const runOnPredicate = (
  predicate: boolean,
  action: Function,
): VoidFunction | (() => VoidFunction) => (): void | VoidFunction =>
  predicate ? action() : undefined;

/**
 * Returns a redirect component for a given round and match.
 * @param {string} roundId The ID of the round to redirect to.
 * @param {string} matchId The ID of the match to redirect to.
 * @return {JSX.Element} The redirect element.
 */
export const getMatchRedirect = (roundId?: string, matchId?: string) => (
  <Redirect to={`/round/${roundId}/match/${matchId}`} />
);

/**
 * Returns a HOC with the given component wrapped by a MobX Observer component.
 * @param {ComponentType<T>} Component The component that should be under an observer.
 * @returns {(props: T) => JSX.Element} A HOC that wraps the given component under an observer.
 */
export const createObserver = <T extends {}>(
  Component: ComponentType<T>,
): ((props: T) => JSX.Element) => (props: T): JSX.Element => (
  <Observer>{() => <Component {...props} />}</Observer>
);
