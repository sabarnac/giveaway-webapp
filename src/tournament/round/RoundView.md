Usage:

```tsx
import Config from "./store/config/Config";
import Round from "./store/round/Round";
import Participant from "./store/round/match/participant/Participant";

const round: Round = new Round(Config.instance, [
  new Participant("FOO"),
  new Participant("BAR")
]);
<RoundView
  matchId="match-1"
  round={round}
  onRoundComplete={() => console.log("Round View Animation Completed!")}
/>;
```
