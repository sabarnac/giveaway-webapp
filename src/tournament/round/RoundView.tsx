import React, { Component, Fragment } from "react";
import { observer, inject } from "mobx-react";
import "./RoundView.scss";
import classNames from "classnames";
import Round from "../../store/round/Round";
import Match from "../../store/round/match/Match";
import MatchView from "./match/MatchView";
import { Redirect } from "react-router";
import { CSSTransition } from "react-transition-group";
import { isInRange } from "../../util/index";
import Config from "../../store/config/Config";

/**
 * Properties of the round view React component.
 */
interface RoundViewProps {
  /**
   * @type {Config} The application config.
   */
  config?: Config;
  /**
   * @type {string} The ID of the current match.
   */
  matchId: string;
  /**
   * @type {Round} The details of the current round.
   */
  round: Round;
  /**
   * @type {() => void} Action to call when the view has finished showing the round.
   */
  onRoundComplete: () => void;
}

/**
 * State of the round view React component.
 */
interface RoundViewState {
  /**
   * @type {number} The current state of the round component (for animations).
   */
  currentState: number;
}

/**
 * React component for the round view.
 */
@inject("config")
@observer
export default class RoundView extends Component<
  RoundViewProps,
  RoundViewState
> {
  /**
   * @type {boolean} Whether the component is mounted or not.
   */
  private _isMounted: boolean = false;

  /**
   * Moves the component to the next animation state.
   */
  private goToNextState = (): void =>
    this._isMounted
      ? this.setState({ currentState: this.state.currentState + 1 })
      : undefined;

  /**
   * Calls the action for round completion, after a delay of 500 (ignoring speed multiplier).
   */
  private onRoundCompleteWithDelay = (): unknown =>
    setTimeout(this.props.onRoundComplete, 500 / this.props.config!.speed);

  /**
   * Returns the match list of the current round.
   * @return {JSX.Element} The matches list view.
   */
  private getMatchList = (): JSX.Element => (
    <div className={classNames("round__list")}>{this.getMatches()}</div>
  );

  /**
   * Returns whether the given match is the current match.
   * @param {Match} match The match to check.
   * @return {boolean} Whether the given match is the current match.
   */
  private isCurrentMatch = (match: Match): boolean =>
    match.id === this.props.matchId;

  /**
   * Returns the index of current match in the tournament.
   * @return {Match} The index of the current match.
   */
  private getCurrentMatchIndex = (): number =>
    this.props.round.matches.findIndex(this.isCurrentMatch);

  /**
   * Returns whether the given match is not a future match.
   * @param {Match} match The match to check.
   * @return {boolean} Whether the given match is not a future match or not.
   */
  private isNotFutureMatch = (_: Match, index: number): boolean =>
    index <= this.getCurrentMatchIndex();

  /**
   * Returns the view for the given match.
   * @param {Match} match The match details.
   * @return {JSX.Element} The match view.
   */
  private getMatchView = (match: Match): JSX.Element => (
    <MatchView
      key={match.id}
      match={match}
      isCurrentMatch={this.isCurrentMatch(match)}
      onMatchComplete={this.goToNextState}
    />
  );

  /**
   * Returns the list of views of all matches of the current round.
   * @return {JSX.Element[]} The list of match views.
   */
  private getMatches = (): JSX.Element[] =>
    this.props.round.matches
      .filter(this.isNotFutureMatch)
      .map(this.getMatchView);

  /**
   * Returns the redirect for the first match if required.
   * @return {JSX.Element} The redirect, or null if it is not required.
   */
  private getFirstMatchRedirect = (): JSX.Element | null =>
    this.getCurrentMatchIndex() === -1 ? (
      <Redirect
        to={`/round/${this.props.round.id}/match/${
          this.props.round.firstMatch.id
        }`}
      />
    ) : null;

  /**
   * Returns whether the next match redirect should be added to the view.
   * @return {boolean} Whether the redirect should be shown.
   */
  private shouldAddNextMatchRedirect = (): boolean =>
    isInRange(
      this.state.currentState,
      0,
      (this.props.round.matches.length - 1) * 2 + 1
    ) &&
    this.state.currentState % 2 === 0 &&
    this.props.round.matches[this.state.currentState / 2].id !==
      this.props.matchId;

  /**
   * Returns the redirect for the next match if required.
   * @return {JSX.Element} The redirect, or null if it is not required.
   */
  private getNextMatchRedirect = (): JSX.Element | null =>
    this.shouldAddNextMatchRedirect() ? (
      <Redirect
        to={`/round/${this.props.round.id}/match/${
          this.props.round.matches[this.state.currentState / 2].id
        }`}
      />
    ) : null;

  /**
   * Returns the details of the current round, wrapped in an animation transition component.
   * @return {JSX.Element} The current round view.
   */
  private getRoundDetails = (): JSX.Element => (
    <CSSTransition
      in={
        isInRange(
          this.state.currentState,
          1,
          this.props.round.matches.length * 2 - 1
        ) && this.getCurrentMatchIndex() !== -1
      }
      timeout={500 / this.props.config!.speed}
      classNames={{
        enter: "",
        enterActive: "round--entering",
        enterDone: "round--entered",
        exit: "",
        exitActive: "round--exiting",
        exitDone: "round--exited"
      }}
      mountOnEnter={true}
      unmountOnExit={true}
      onExited={this.onRoundCompleteWithDelay}
    >
      <div
        className={classNames("round")}
        style={{
          transition: `opacity ${500 / this.props.config!.speed}ms ease-in-out`
        }}
      >
        {this.getMatchList()}
      </div>
    </CSSTransition>
  );

  /**
   * Lifecycle method that runs before component mount.
   */
  public componentWillMount = (): void => {
    this.setState({
      currentState:
        this.getCurrentMatchIndex() === -1 ? 0 : this.getCurrentMatchIndex() * 2
    });
  };

  /**
   * Lifecycle method that runs after component mount.
   */
  public componentDidMount = (): void => {
    this._isMounted = true;
    this.goToNextState();
  };

  /**
   * Lifecycle method that runs after component update.
   */
  public componentDidUpdate = (prevProps: RoundViewProps): void => {
    if (this.props.matchId !== prevProps.matchId) {
      this.setState({ currentState: this.getCurrentMatchIndex() * 2 + 1 });
    }
  };

  /**
   * Lifecycle method that runs before component unmount.
   */
  public componentWillUnmount = (): void => {
    this._isMounted = false;
  };

  /**
   * Renders the component.
   * @return {JSX.Element} The rendered component.
   */
  public render = (): JSX.Element => (
    <Fragment
      key={`round-${this.props.round.id}-match-${this.getCurrentMatchIndex()}`}
    >
      {this.getRoundDetails()}
      {this.getFirstMatchRedirect()}
      {this.getNextMatchRedirect()}
    </Fragment>
  );
}
