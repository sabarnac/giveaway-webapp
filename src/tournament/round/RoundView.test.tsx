import React from "react";
import RoundView from "./RoundView";
import Round from "../../store/round/Round";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyRound } from "../../util/test";

describe("RoundView Component.", () => {
  it("Roundes snapshot.", () => {
    const round: Round = createDummyRound();
    const component: ReactTestRenderer = create(
      <RoundView
        round={round}
        matchId={round.firstMatch.id}
        onRoundComplete={() => {}}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
