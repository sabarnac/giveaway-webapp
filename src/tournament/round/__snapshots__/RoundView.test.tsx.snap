// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoundView Component. Roundes snapshot. 1`] = `
<div
  classNames={
    Object {
      "enter": "",
      "enterActive": "round--entering",
      "enterDone": "round--entered",
      "exit": "",
      "exitActive": "round--exiting",
      "exitDone": "round--exited",
    }
  }
  in={true}
  mountOnEnter={true}
  onExited={[Function]}
  timeout={500}
  unmountOnExit={true}
>
  <div
    className="round"
    style={
      Object {
        "transition": "opacity 500ms ease-in-out",
      }
    }
  >
    <h2>
      Round 1
    </h2>
    <div
      className="round__list"
    >
      <div
        classNames={
          Object {
            "enter": "",
            "enterActive": "match--entering",
            "enterDone": "match--entered",
            "exit": "",
            "exitActive": "match--exiting",
            "exitDone": "match--exited",
          }
        }
        in={true}
        mountOnEnter={true}
        onEntering={[Function]}
        onExited={[Function]}
        timeout={500}
        unmountOnExit={true}
      >
        <div
          className="match"
          style={
            Object {
              "transition": "opacity 500ms ease-in-out",
            }
          }
        >
          <div
            className="match__list"
          >
            <div
              className="participant participant-entry"
            >
              <div
                className="participant-entry__avatar"
              >
                <img
                  alt="barfoo-6"
                  src="foobar-6"
                />
              </div>
              <div
                className="participant-entry__name"
              >
                <strong>
                  Fozbaz 6
                </strong>
              </div>
            </div>
            <div
              className="participant participant-entry"
            >
              <div
                className="participant-entry__avatar"
              >
                <img
                  alt="barfoo-4"
                  src="foobar-4"
                />
              </div>
              <div
                className="participant-entry__name"
              >
                <strong>
                  Fozbaz 4
                </strong>
              </div>
            </div>
          </div>
          <div
            classNames={
              Object {
                "enter": "",
                "enterActive": "match__interim--entering",
                "enterDone": "match__interim--entered",
                "exit": "",
                "exitActive": "match__interim--exiting",
                "exitDone": "match__interim--exited",
              }
            }
            in={false}
            mountOnEnter={true}
            onEntered={[Function]}
            onExited={[Function]}
            timeout={200}
            unmountOnExit={true}
          >
            <div
              className="match__interim"
              style={
                Object {
                  "transition": "opacity 200ms ease-in-out",
                }
              }
            >
              <div
                className="participant participant-entry"
              >
                <div
                  className="participant-entry__avatar"
                >
                  <img
                    alt="barfoo-6"
                    src="foobar-6"
                  />
                </div>
                <div
                  className="participant-entry__name"
                >
                  <strong>
                    Fozbaz 6
                  </strong>
                </div>
              </div>
            </div>
          </div>
          <div
            classNames={
              Object {
                "enter": "",
                "enterActive": "match__interim--entering",
                "enterDone": "match__interim--entered",
                "exit": "",
                "exitActive": "match__interim--exiting",
                "exitDone": "match__interim--exited",
              }
            }
            in={false}
            mountOnEnter={true}
            onEntered={[Function]}
            onExited={[Function]}
            timeout={200}
            unmountOnExit={true}
          >
            <div
              className="match__interim"
              style={
                Object {
                  "transition": "opacity 200ms ease-in-out",
                }
              }
            >
              <div
                className="participant participant-entry"
              >
                <div
                  className="participant-entry__avatar"
                >
                  <img
                    alt="barfoo-4"
                    src="foobar-4"
                  />
                </div>
                <div
                  className="participant-entry__name"
                >
                  <strong>
                    Fozbaz 4
                  </strong>
                </div>
              </div>
            </div>
          </div>
          <div
            classNames={
              Object {
                "enter": "",
                "enterActive": "match__winner--entering",
                "enterDone": "match__winner--entered",
                "exit": "",
                "exitActive": "match__winner--exiting",
                "exitDone": "match__winner--exited",
              }
            }
            in={false}
            mountOnEnter={true}
            onEntered={[Function]}
            timeout={200}
            unmountOnExit={true}
          >
            <div
              className="match__winner"
              style={
                Object {
                  "transition": "opacity 200ms ease-in-out",
                }
              }
            >
              <div
                className="participant participant-entry"
              >
                <div
                  className="participant-entry__avatar"
                >
                  <img
                    alt="barfoo-4"
                    src="foobar-4"
                  />
                </div>
                <div
                  className="participant-entry__name"
                >
                  <strong>
                    Fozbaz 4
                  </strong>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
